### PR TITLE
Support `http/json` encoded requests and responses for `otlpmetrichttp` client

### LINKED ISSUE #7803: Support `http/json` in `otlpmetrichttp`
### Problem Statement

`otlpmetrichttp` does not support `http/json` encoding format.

### Proposed Solution

Add support for encoding in protojson format and make it configurable, keeping proto as the default option.

### Additional Context

https://github.com/open-telemetry/opentelemetry-go-contrib/issues/6339
https://github.com/open-telemetry/opentelemetry-go-contrib/issues/8399

<sub>**Tip**: [React](https://github.blog/news-insights/product-news/add-reactions-to-pull-requests-issues-and-comments/) with 👍 to help prioritize this issue. Please use comments to provide useful context, avoiding `+1` or `me too`, to help us triage it. Learn more [here](https://opentelemetry.io/community/end-user/issue-participation/).</sub>

### PR DESCRIPTION
**Pull request by petern48**
Add support for encoding in protojson format and make it configurable, keeping proto as the default option.

closes #7803

This is needed for the following:

https://github.com/open-telemetry/opentelemetry-go-contrib/issues/6339
https://github.com/open-telemetry/opentelemetry-go-contrib/issues/8399

### CODE REVIEW

**Comment by petern48** [comment]
The `check-links` job is failing due to links that were already in the `.tmpl` files.
https://github.com/open-telemetry/opentelemetry-go/actions/runs/21115879038/job/60721331105?pr=7804

The failing [codecov CI](https://github.com/open-telemetry/opentelemetry-go/pull/7804/checks?check_run_id=60721352548) makes sense to me as well. The modifications to `collector.go` are tested (and needed) in `client_test.go`, but codecov just doesn't detect that.

I'm just going to leave the PR here and await feedback.

**Comment by marcschaeferger** [comment]
@petern48 Are you working on it?

File: CHANGELOG.md [hunk 1]
```diff
@@ -16,6 +16,7 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
 - Add `go.opentelemetry.io/otel/semconv/v1.39.0` package.
   The package contains semantic conventions from the `v1.39.0` version of the OpenTelemetry Semantic Conventions.
   See the [migration documentation](./semconv/v1.39.0/MIGRATION.md) for information on how to upgrade from `go.opentelemetry.io/otel/semconv/v1.38.0.`(#7783)
+- Added `http/json` support to `go.opentelemetry.io/otel/exporters/otlp/otlpmetric/otlpmetrichttp` (#7804)
```
Comment (dmathieu) [inline CHANGELOG.md]:
```suggestion
- Added `http/json` support, and the `WithEncoding` option to `go.opentelemetry.io/otel/exporters/otlp/otlpmetric/otlpmetrichttp`. (#7804)
```

File: exporters/otlp/otlpmetric/otlpmetrichttp/config.go [hunk 2]
```diff
@@ -109,6 +119,13 @@ func WithCompression(compression Compression) Option {
 	return wrappedOption{oconf.WithCompression(oconf.Compression(compression))}
 }
 
+// WithEncoding sets the encoding the Exporter will use to encode the payload.
+//
+// By default, if this option is not passed, protobuf encoding will be used.
```
Comment (dmathieu) [inline exporters/otlp/otlpmetric/otlpmetrichttp/config.go]:
```suggestion
// By default, if this option is not used, or an invalid value is passed, the protobuf encoding will be used.
```

**Review by dmathieu — COMMENTED** [review]
We will need to do this for other signals as well.

**Comment by marcschaeferger** [comment]
I'm happy to work on it, if @petern48 is not working on it anymore. I'll look back in 7 - 14 days

**Review by dmathieu — APPROVED** [review]

**Review by flc1125 — APPROVED** [review]
From the perspective of the request side, there are currently no issues.

However, I'd like to confirm a question about the protocol: the current PR has resolved the issue of the request side's support for `http/json`, but do we still need to ensure compatibility with the response format returned by the other server?

Or, if necessary, we can submit a separate PR to handle this issue.

cc @open-telemetry/go-maintainers

**Comment by petern48** [comment]
@flc1125 Thanks for the catch. I've implemented json encoding support for the response side in https://github.com/open-telemetry/opentelemetry-go/pull/7804/changes/c9eaf4a6df05bb55edd8f235b820393b279575d9. Let me know what you think.

Note: I followed the same behavior as the collector (https://github.com/open-telemetry/opentelemetry-collector/pull/7451/) using `mime.ParseMediaType()` to handle headers that include optional parameters like `charset` (an example is shown in the test).

**Review by flc1125 — APPROVED** [review]

**Comment by petern48** [comment]
Gentle bump. Looking for a third approval.

File: CHANGELOG.md [hunk 3]
```diff
@@ -32,6 +32,7 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
 - Add support for per-series start time tracking for cumulative metrics in `go.opentelemetry.io/otel/sdk/metric`.
   Set `OTEL_GO_X_PER_SERIES_START_TIMESTAMPS=true` to enable. (#8060)
 - Add `WithCardinalityLimitSelector` for metric reader for configuring cardinality limits specific to the instrument kind. (#7855)
+- Add `http/json` support, and the `WithEncoding` option to `go.opentelemetry.io/otel/exporters/otlp/otlpmetric/otlpmetrichttp`. (#7804)
```
Comment (dmathieu) [inline CHANGELOG.md]: The merge moved this into the released section. Could you move it to unreleased?

File: internal/shared/otlp/otlpmetric/oconf/optiontypes.go.tmpl [hunk 4]
```diff
@@ -31,6 +31,16 @@ const (
...
 )
 
+// Encoding describes the encoding used for payloads sent to the collector.
+type Encoding int
+
+const (
+	// ProtobufEncoding tells the driver to send payloads using protobuf encoding.
+	ProtobufEncoding Encoding = iota
+	// JSONEncoding tells the driver to send payloads using JSON encoding.
+	JSONEncoding
```
Comment (pellared) [inline internal/shared/otlp/otlpmetric/oconf/optiontypes.go.tmpl]: Ths does not seem to follow Go conventions e.g. https://pkg.go.dev/net/http#ConnState

This should rather be

```suggestion
	// EncodingProtobuf tells the driver to send payloads using protobuf encoding.
	EncodingProtobuf Encoding = iota
	// EncodingJSON tells the driver to send payloads using JSON encoding.
	EncodingJSON
```

Comment (petern48) [reply internal/shared/otlp/otlpmetric/oconf/optiontypes.go.tmpl]: Done! Thanks for the catch @pellared

**Comment by petern48** [comment]
Was there anything else you were looking for to earn your approval, @pellared? Noticed you merged main, but no follow-up feedback.

File: exporters/otlp/otlpmetric/otlpmetrichttp/client.go [hunk 5]
```diff
@@ -142,7 +156,15 @@ func (c *client) UploadMetrics(ctx context.Context, protoMetrics *metricpb.Resou
 	pbRequest := &colmetricpb.ExportMetricsServiceRequest{
 		ResourceMetrics: []*metricpb.ResourceMetrics{protoMetrics},
 	}
-	body, err := proto.Marshal(pbRequest)
+
+	var body []byte
+	var err error
+	switch c.encoding {
+	case EncodingJSON:
+		body, err = protojson.Marshal(pbRequest)
```
Comment (MrAlias) [inline exporters/otlp/otlpmetric/otlpmetrichttp/client.go]: This needs OTLP-specific marshal options. OTLP/JSON requires enum fields to be encoded as integer values, but default `protojson.Marshal` can emit enum names instead.

That means a non-empty metrics payload can be serialized into JSON that is valid protobuf JSON but not valid OTLP/JSON. I think this should use `protojson.MarshalOptions{UseEnumNumbers: true}`.

It would also be good to cover this with a test that exports a non-empty metrics payload and checks the raw JSON for numeric enum values, e.g. `aggregationTemporality`.

File: exporters/otlp/otlpmetric/otlpmetrichttp/client.go [hunk 6]
```diff
@@ -192,19 +214,30 @@ func (c *client) UploadMetrics(ctx context.Context, protoMetrics *metricpb.Resou
...
+			mediatype, _, err := mime.ParseMediaType(resp.Header.Get("Content-Type"))
+			if err != nil {
+				mediatype = ""
+			}
+			var respProto colmetricpb.ExportMetricsServiceResponse
+			switch mediatype {
+			case "application/x-protobuf":
+				err = proto.Unmarshal(respData.Bytes(), &respProto)
+			case "application/json":
+				err = protojson.Unmarshal(respData.Bytes(), &respProto)
```
Comment (MrAlias) [inline exporters/otlp/otlpmetric/otlpmetrichttp/client.go]: OTLP/JSON receivers are supposed to ignore unknown fields for forward compatibility, but default `protojson.Unmarshal` rejects them.

That makes the client stricter than the OTLP spec when parsing JSON responses. I think this should use `protojson.UnmarshalOptions{DiscardUnknown: true}` so a newer server can add fields without breaking older clients.

File: exporters/otlp/otlpmetric/otlpmetrichttp/internal/otest/collector.go [hunk 7]
```diff
@@ -311,7 +312,12 @@ func (c *HTTPCollector) record(r *http.Request) ExportResult {
 		return ExportResult{Err: err}
 	}
 	pbRequest := &collpb.ExportMetricsServiceRequest{}
-	err = proto.Unmarshal(body, pbRequest)
+	switch contentType {
+	case "application/json":
+		err = protojson.Unmarshal(body, pbRequest)
```
Comment (MrAlias) [inline exporters/otlp/otlpmetric/otlpmetrichttp/internal/otest/collector.go]: Same issue in the test collector: OTLP/JSON receivers should ignore unknown fields, but default `protojson.Unmarshal` rejects them.

Since this helper is standing in for an OTLP/HTTP receiver in tests, I think it should also use `protojson.UnmarshalOptions{DiscardUnknown: true}`.

File: exporters/otlp/otlpmetric/otlpmetrichttp/client_test.go [hunk 8]
```diff
@@ -76,6 +79,100 @@ func TestClientWithHTTPCollectorRespondingPlainText(t *testing.T) {
...
+	coll, err := otest.NewHTTPCollector("", nil)
+	require.NoError(t, err)
+
+	addr := coll.Addr().String()
+	opts := []Option{WithEndpoint(addr), WithInsecure(), WithEncoding(EncodingJSON)}
+	cfg := oconf.NewHTTPConfig(asHTTPOptions(opts)...)
+	client, err := newClient(cfg)
+	require.NoError(t, err)
+
+	require.NoError(t, client.UploadMetrics(ctx, &mpb.ResourceMetrics{}))
```
Comment (MrAlias) [inline exporters/otlp/otlpmetric/otlpmetrichttp/client_test.go]: These tests don’t exercise the spec-sensitive parts of the JSON path yet. Right now they only send empty `ResourceMetrics` payloads or round-trip JSON produced by the same default `protojson` implementation, so they would still pass if we emitted enum names or rejected unknown fields.

I think we need a couple of more targeted cases here:
- a request-side test that inspects the raw JSON body for a non-empty metrics payload and proves enum values are encoded numerically
- a response-side test that injects an unknown JSON field and verifies the client still accepts the response
- ideally a collector/helper test for the same unknown-field behavior on the receive side

**Review by MrAlias — CHANGES_REQUESTED** [review]
Thanks for pushing this forward. I checked the JSON path against the OTLP/HTTP spec and found a couple of places where we’re still implementing generic protobuf JSON rather than OTLP/JSON.

File: exporters/otlp/otlpmetric/otlpmetrichttp/client_test.go (hunk 8 above)
Comment (petern48) [reply exporters/otlp/otlpmetric/otlpmetrichttp/client_test.go]: I added tests for the first two. It felt odd to add a test for the third one since there aren't already any existing tests for the collector anywhere. If you feel strongly about it, I can create a new file `collector_test.go` and add it there.

File: exporters/otlp/otlpmetric/otlpmetrichttp/client.go [hunk 9]
```diff
@@ -152,7 +178,15 @@ func (c *client) UploadMetrics(ctx context.Context, protoMetrics *metricpb.Resou
 	pbRequest := &colmetricpb.ExportMetricsServiceRequest{
 		ResourceMetrics: []*metricpb.ResourceMetrics{protoMetrics},
 	}
-	body, err := proto.Marshal(pbRequest)
+
+	var body []byte
+	var err error
+	switch c.encoding {
+	case EncodingJSON:
+		body, err = otlpProtoJSON.Marshal(pbRequest)
```
Comment (MrAlias) [inline exporters/otlp/otlpmetric/otlpmetrichttp/client.go:186]: This still uses the generic protobuf JSON mapping for byte fields. Metric exemplars carry `traceId` and `spanId` as byte fields, so `protojson` emits them as base64 even though [OTLP/JSON requires hexadecimal IDs](https://opentelemetry.io/docs/specs/otlp/#json-protobuf-encoding). A conforming receiver can reject the whole request when an exported data point has a trace exemplar.

Could we use an OTLP-aware encoding path here, update the test collector's matching decoder, and add a raw-body test with non-empty exemplar IDs?

**Review by MrAlias — COMMENTED** [review]
The earlier enum and unknown-field issues are addressed. There is still one OTLP/JSON wire-format gap in the request path.

**Comment by MrAlias** [comment]
@petern48 are you still able to work on this? Should be close this for now?

**Comment by marcschaeferger** [comment]
If not, I can take over
